---
output: 
    github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
# Banksy

<!-- badges: start -->
<!-- badges: end -->

Banksy is an R package that incorporates spatial information to cluster cells 
in a feature space (e.g. gene expression). Spatial information is incorporated 
by averaging the features of the k nearest neighbours to generate new 
'neighbour' features for a given cell. This is concatenated to the cell's own 
features to generate a combined feature matrix which is used for constructing a 
nearest neighbour network. Leiden clustering is used to obtain 
spatially-informed clusters.

# Installation

```{r, eval=F}
# Install from GitHub
remotes::install_github("jleechung/Banksy")

# Install from Bioconductor (sooon)
```

# Documentation

NOTE: These links will not work at the moment since the repo is still private.
Refer to the [vignettes](https://github.com/jleechung/Banksy/tree/main/vignettes) for the time being.

*Banksy* comes installed with [documentation](https://jleechung.github.io/Banksy/reference/index.html) 
of main functions and their usage, along with several vignettes which detail 
different use cases:

* [Working with Banksy objects](https://jleechung.github.io/Banksy/articles/banksy-object.html):
  Introduction to the *BanksyObject* class which serves as a container for *Banksy*.  
  
* [Finding optimal clustering parameters](https://jleechung.github.io/Banksy/articles/grid-search.html):
  Illustrates a grid search of parameters which best cluster cells.  

* [Analysing multiple datasets](https://jleechung.github.io/Banksy/articles/multiple-datasets.html):
  Illustrates analysis of multiple spatial transcriptomic datasets.

* [Integrating *Banksy* with other packages](https://jleechung.github.io/Banksy/articles/integrating-pkgs.html):
  Illustrates how users can integrate *Banksy* with packages like *Seurat* and *Giotto*.

# Basic Usage

## Input data

Inputs consist of an expression matrix and cell locations. Sample data is 
provided with the package:
```{r}
library(Banksy)

expr <- readRDS(system.file('/extdata/expression.rds', package = 'Banksy'))
locs <- readRDS(system.file('/extdata/locations.rds', package = 'Banksy'))
```

The gene expression matrix for cells should be a `matrix`:
```{r}
class(expr)
head(expr[,1:5])
```

while cell locations should be supplied as a `data.frame`:
```{r}
class(locs)
head(locs)
```

## Spatial clustering

First, create a *BanksyObject* with the expression matrix and cell locations.

```{r}
bank <- BanksyObject(own.expr = expr,
                     cell.locs = locs)
```

Next, normalize the expression matrix, compute the neighbour matrix, and scale 
the resulting gene-cell and neighbour feature-cell matrix.

```{r}
bank <- NormalizeBanksy(bank, normFactor = 100)
bank <- ComputeBanksy(bank)
bank <- ScaleBanksy(bank)
```

At this point, the joint expression matrix (gene-cell matrix and neighbour 
feature-cell matrix) can be extracted with *getBanksyMatrix*, which returns
the joint matrix and cell locations:

```{r}
joint <- getBanksyMatrix(bank, lambda = 0.25)

joint$expr[1:5,1:5]
joint$locs[1:5,]
```

Next, we obtain cluster assignments for the following parameters:

* `lambda`. A mixing parameter from 0 to 1 which determines how much spatial 
   information is incorporated.  
* `resolution`. Leiden clustering resolution.  
* `kneighbours`. Number of k neighbours to use for constructing sNN.

```{r}
set.seed(1234)
bank <- ClusterBanksy(bank, lambda = 0.5,
                            resolution = 0.8,
                            kneighbours = 40)
```

## Visualization

UMAP visualization:
```{r}
plotUMAP(bank, by = 'res0.8_lam0.5_k40',  reduction = 'umap_0.5', pt.size = 0.02, main.size = 10)
```

Spatial plot:
```{r}
plotSpatialDims(bank, by = 'res0.8_lam0.5_k40', pt.size = 0.8, main.size = 10)
```

## Subsetting

The *BanksyObject* can be subset by cells, features or metadata columns.
```{r}
bankSubset1 <- SubsetBanksy(bank,
                            dims = (sdimx > 1000 & sdimx < 4000) & 
                                   (sdimy > 2000 & sdimy < 3000),
                            cells = c('cell_364', 'cell_367', 'cell_384','cell_389'),
                            features = sample(rownames(own.expr(bank)), 10))
head(bankSubset1)
```

```{r}
bankSubset2 <- SubsetBanksy(bank,
                            metadata = res0.8_lam0.5_k40 %in% c(3,6))
plotSpatialDims(bankSubset2, by = 'res0.8_lam0.5_k40', 
                pt.size = 0.5, main.size = 10)
```

# Session information
```{r}
sessionInfo()
```

## To-dos

* Pass devtools::check: S3/4 documentation
* Normalization methods for multiple datasets
* Clustering methods
* Optimization for handling large matrices (Rcpp)
* Subsetting and Connecting clusters for multiple datasets
* Unit tests
