---
title: "Analysing multiple datasets"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analysing multiple datasets}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

*Banksy* allows the analysis of multiple datasets. Apart from taking in single 
inputs, the *BanksyObject* constructor can take in a list of expression matrices
and cell locations. The user should ensure that there is one-one correspondence 
between the gene-cell matrix and cell locations.

Three spatial transcriptomic datasets comprising single-cell sequencing of brain
organoids are attached with the *Banksy* package, and can be loaded as follows: 

```{r setup, eval=T}
library(Banksy)
library(gridExtra)

gcm1 <- readRDS(system.file('/extdata/20200221-B4-40-SCT-gcm.rds', package = 'Banksy'))
gcm2 <- readRDS(system.file('/extdata/20200627-B4-40-SCT-gcm.rds', package = 'Banksy'))
gcm3 <- readRDS(system.file('/extdata/20200309-B4-220-SCT-gcm.rds', package = 'Banksy'))

loc1 <- readRDS(system.file('/extdata/20200221-B4-40-SCT-loc.rds', package = 'Banksy'))
loc2 <- readRDS(system.file('/extdata/20200627-B4-40-SCT-loc.rds', package = 'Banksy'))
loc3 <- readRDS(system.file('/extdata/20200309-B4-220-SCT-loc.rds', package = 'Banksy'))
```

Each dataset comprises 1200 cells, and all have been pre-normalized to reduce 
batch effects.

## Constructing *BanksyObject* for multiple datasets

Here we construct the *BanksyObject*. For ease of reference, names can be 
provided to each dataset. For consistency across datasets, the gene set for 
analysis can either be the intersection or union of genes across all 
datasets, controlled by the `genes.filter` argument. If set to `union`, 
missing genes in datasets will be imputed with zeroes. Additionally, genes can 
be filtered by the minimum number of cells expressing those genes, controlled
by the `min.cells.expressed` argument. This is only active when `genes.filter` 
is set to `intersect`. Here, we take the union of all genes.

```{r construct, eval=T}
dnames <- c('d1', 'd2', 'd3')
gcm <- list(gcm1, gcm2, gcm3)
loc <- list(loc1, loc2, loc3)
names(gcm) <- names(loc) <- dnames

bank <- BanksyObject(own.expr = gcm,
                     cell.locs = loc,
                     genes.filter = 'union')
bank
```

The *BanksyObject* here contains three assays, each containing 1200 organoids, 
with 561 genes. Each dataset contains two spatial dimensions.

The rest of the workflow is similar to the single dataset case. We compute
the neighbour-feature cell matrix and scale the datasets separately (No 
normalization step here is required as the data has been pre-normalized).

```{r compute, eval=T}
bank <- ComputeBanksy(bank)
bank <- ScaleBanksy(bank)
```

One can obtain the combined joint matrix by calling *getBanksyMatrix*:
```{r joint, eval=T}
combined <- getBanksyMatrix(bank)
lapply(combined, dim)
```

This concatenates the gene-cell matrices and neighbour-feature cell matrices
across all datasets.

## Clustering and visualisation

Next, we cluster the organoids from all datasets. A smaller resolution is used
here, given the relatively limited number of organoids per dataset.

```{r cluster, eval=T}
bank <- ClusterBanksy(bank,
                      resolution = 0.8,
                      lambda = 0.5,
                      kneighbours = 40,
                      verbose = TRUE)
```

We can visualise all spatial dimension plots across the three datasets with
*grid.arrange* from the *gridExtra* package. The *plotSpatialDims* function
can be called with the `dataset` argument to plot spatial dimensions for
a specifed dataset. Here, we iterate over all datasets and generate
spatial dimension plots. UMAP visualisation is similar to the one dataset case.

```{r spatdim, eval=T, fig.width = 8, fig.height = 7, fig.align='center'}
## UMAP
params <- 'res0.8_lam0.5_k40'
umap <- plotUMAP(bank, params, pt.size = 0.02, main.size = 10)

## Spatial plots
dsetnames <- names(own.expr(bank))
spatdims <- lapply(dsetnames,
                   function(x) plotSpatialDims(bank, params, dataset = x,
                                               pt.size = 0.9, main.size = 10))
                                               
plots <- c(list(umap), spatdims)

do.call(grid.arrange, c(plots, ncol=2))
```

(Note: *ConnectClusters* and *SubsetBanksy* are currently not supported for 
multiple datasets)

## Session Information
```{r sess}
sessionInfo()
```

