---
title: "Mouse Brain Organoid MERFISH data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Mouse Brain Organoid MERFISH data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

*BANKSY* allows the analysis of multiple datasets. Here, we demonstrate this 
functionality with a MERFISH mouse brain organoid dataset. Three spatial transcriptomic datasets comprising single-cell sequencing of brain
organoids are attached with the package. See `?Banksy::organoid` for more 
details. 

```{r setup, eval=T}
library(Banksy)
library(gridExtra)

data("organoid")
expression <- organoid$expression
locations <- organoid$locations
```

Each dataset comprises 1200 cells, and all have been pre-normalized to reduce 
batch effects.

## Constructing *BanksyObject* for multiple datasets

Here we construct the *BanksyObject*. Apart from taking in single inputs, the *BanksyObject* constructor can take in a list of expression matrices and cell locations. The user should ensure that there is one-one correspondence 
between the gene-cell matrix and cell locations.

For ease of reference, names can be provided to each dataset. For consistency 
across datasets, the gene set for analysis can either be the intersection or 
union of genes across all datasets, controlled by the `genes.filter` argument. 
If set to `union`, missing genes in datasets will be imputed with zeroes. Additionally, genes can be filtered by the minimum number of cells expressing 
those genes, controlled by the `min.cells.expressed` argument. This is only 
active when `genes.filter` is set to `intersect`. Here, we take the intersection
of genes.

```{r construct, eval=T}
dnames <- c('d1', 'd2', 'd3')
names(expression) <- names(locations) <- dnames

bank <- BanksyObject(own.expr = expression,
                     cell.locs = locations,
                     genes.filter = 'intersect')
bank
```

The *BanksyObject* here contains three assays, each containing 1200 organoids, 
with 561 genes. Each dataset contains two spatial dimensions.

The rest of the workflow is similar to the single dataset case. We compute
the neighbor-augmented matrix and scale the datasets separately (no 
normalization step here is required as the data has been pre-normalized).

```{r compute, eval=T}
bank <- ComputeBanksy(bank)
bank <- ScaleBanksy(bank)
```

One can obtain the combined joint matrix by calling *getBanksyMatrix*. This 
concatenates the gene-cell matrices and neighbor-feature cell matrices across 
all datasets.

```{r joint, eval=T}
combined <- getBanksyMatrix(bank)
lapply(combined, dim)
```

## Clustering and visualisation

Next, we cluster the organoids from all datasets. In order to find regions of 
mis-differentiated and necrotic core tissue, we run BANKSY in zone-finding
mode with `lambda=0.75`.

```{r cluster, eval=T}
bank <- RunPCA(bank, lambda = 0.75, npcs = 30)

bank <- ClusterBanksy(bank, lambda=0.75, pca=TRUE, npcs=30,
                      method='leiden', k.neighbors=150, resolution=1)

```

Visualise zones in each of the three datasets:

```{r spatdim, eval=T, fig.width=9, fig.height=7, fig.align='center', warning=F}
plist <- lapply(dnames, function(d) {
  plotSpatial(bank, dataset = d, by = clust.names(bank), type = 'discrete', pt.size = 1)
})

grid.arrange(grobs = plist, nrow = 2)
```

## Session Information
```{r sess}
sessionInfo()
```

